add global pallette (#2)
ritza-co/coding-with-replit: manuscript/10-productivity-hacks.md

Similarly to learning to touch type, there is often a steep *learning curve* when you start to use advanced code editing features. They might even substantially slow you down at first, but once you master them you'll soar past the limits of what you could achieve without these aids.

## Using the global command palette

If you hit `Ctrl+K` (`Cmd+K` on MacOS) you'll see the following modal pop up, which lets you navigate through different parts of Repl.it at lightning speed using only your keyboard. If you have a lot of files, it's often useful to open them like this rather than scrolling through the directory structure in the files pane (the `find` option searches through files by their *name* while the `search` option searches through files by their *contents*.)

> ![**Image 1:** *Using the global command palette.*](https://codewithrepl.it/img/10-global-command-palette.png)

Reviewer: ```suggestion
![**Image 1:** *Using the global command palette.*](https://www.codewithrepl.it/img/10-global-command-palette.png)
```

Opening up the multiplayer, version control, and settings panes using this method is also faster once the habit is ingrained compared to moving the mouse to the small icons on the left bar. And while pressing `Ctrl+Enter` or `Cmd+Enter` to run your code is faster than choosing "Run" from this global command ette, `Ctrl+K` is only a single shortcut to remember and it will remind you of any other shortcuts you can't recall.

## Using the code editing command palette

The code command palette is similar to the global command palette, but it's specific to editing and navigating your code, allowing you to do advance find+replace procedures, jump to specific functions and more.

To access the code command palette press `F1` or `Ctrl+Shift+P` (`cmd+shift+P` on MacOS). Note that if you're using Firefox the second option will open an incognito window instead.

You can use the shortcuts directly from the command palette by selecting the code you wish to edit and clicking on the command in the drop-down menu, or use it to refresh your memory on the keybindings associated with the shortcuts you use often.

> ![**Image 3:** *Opening the command palette.*](https://codewithrepl.it/img/10-open-cmd-palette.png)

Reviewer: ```suggestion
![**Image 3:** *Opening the command palette.*](resources/10-open-cmd-palette.png)
```

Instead of typing out all three lines, you can type out the first one, leave your cursor position on that line, and press `Shift+Alt+down` (`shift+option+down` on MacOS) twice. This will create two copies of the line, directly below the original one, and then you can simply change the number in the variable to account for the second two balls.

> ![**Image 4:** *Copying the current selected line.*](https://codewithrepl.it/img/10-copy-current-line.gif)

Reviewer: ```suggestion
![**Image 4:** *Copying the current selected line.*](resources/10-copy-current-line.png)
```

As an example, below you can see how we might use this to first delete one of the our `elif` blocks by doing two "delete line" operations. We then change our random speed to be constant by using a "delete to end of line" operation from the `=` sign and then typing our constant.

> ![**Image 5:** *Deleting selected lines of code.*](https://codewithrepl.it/img/10-deleting-lines.gif)

Reviewer: ```suggestion
![**Image 5:** *Deleting selected lines of code.*](resources/10-deleting-lines.png)
```

## Inserting blank lines

It's also common to need to add a new line of code above or below the current one. Instead of using your mouse or arrow keys to get to the right place and then pressing `Enter`, you can instead use an "insert line" operation.

Press `Ctrl+Shift+Enter` (`cmd+shift+enter` on MacOS) to insert a blank line directly above the current one and move the cursor to the start of it (Repl.it will even maintain the current level of indentatin for you).

> ![**Image 6:** *Inserting lines.*](https://codewithrepl.it/img/10-inserting-lines.gif)

Reviewer: ```suggestion
![**Image 6:** *Inserting lines.*](resources/10-inserting-lines.png)
```

Now your code's indentation will be fixed.

> ![**Image 7:** *Indenting a line.*](https://codewithrepl.it/img/10-indent.gif)

Reviewer: ```suggestion
![**Image 7:** *Indenting a line.*](resources/10-indent.png)
```

### Moving blocks of code within a file

Sometimes you need to move a block of code up or down in the file. For example, our `update()` function uses our `move()` function, but `move()` is only defined later. For readability, it's good to try ensure that your functions only call functions that have already been defined further up (assuming that someone else is reading the code top down, they will remember the `move()` function's definition before seeing it used).

Instead of cutting and pasting a block, you can shunt it by pressing `Alt+up` or `Alt+down` (`option+up` and `option+down` on MacOS). As with the others, this works on the line under your cursor or a larger selection.

> ![**Image 8:** *Moving the current line selection.*](https://codewithrepl.it/img/10-move-current-line-selection.gif)

Reviewer: ```suggestion
![**Image 8:** *Moving the current line selection.*](resources/10-move-current-line-selection.png)
```

## Adding cursors

Sometimes it's useful to make exactly the same changes in multiple places at once. For example, we might want to rename our `speed` attribute to `velocity`. Put your cursor anywhere on the word that you want to change and press `Ctrl+D` (`cmd+D` on MacOS). Repeatedly press `Ctrl+D` to select matching words individually, each with their own cursor. Now you can apply edits and they will appear at each selection, as below.

> ![**Image 9:** *Adding cursors to multiple instances of the same selection.*](https://codewithrepl.it/img/10-add-cursor-to-selection.gif)

Reviewer: ```suggestion
![**Image 9:** *Adding cursors to multiple instances of the same selection.*](resources/10-add-cursor-to-selection.png)
```

If you want multiple cursors on consecutive lines, press `Ctrl+Alt+up` or `Ctrl+Alt+down` (`cmd+option+up` and `cmd+option+down` on MacOs). For example, if we want a square game we could change both width and height to be `1000` simultaneously as follows.

> ![**Image 10:** *Adding cursors to multiple lines.*](https://codewithrepl.it/img/10-adding-cursors.gif)

Reviewer: ```suggestion
![**Image 10:** *Adding cursors to multiple lines.*](resources/10-adding-cursors.png)
```

## Navigating to specific pieces of code

Sometimes, especially in larger projects, you'll call a function or instantiate an object far from where that function or object is defined (either thousands of lines away in the same file, or in a different file altogether).

If you're reading a piece of code that calls a function and you want to quickly see what that function actually does, you can use the **go to definition** keybinding (`F12` or `cmd+F12` on MacOS). This will jump to the definition of the function or class selected. The **peek definition** has a similar functionality, but instead of jumping to the definition, it opens in a separate modal. For example, below, the cursor is on the instantiation of `Ball()` and we can quickly see how this class is defined.

> ![**Image 11:** *Peeking the definition.*](https://codewithrepl.it/img/10-peek-definition.gif)

Reviewer: ```suggestion
![**Image 11:** *Peeking the definition.*](resources/10-peek-definition.png)
```

The **go to line** operation (`Ctrl+G`) allows you to navigate to a line by giving its line number. This is useful to track down the source of those error messages that tell you what line had an issue, or if you're on a call with someone who says "I'm looking at line 23" and you can quickly jump to the same place.

Finally, you can open a specific file by searching for a part of the name by pressing `Ctrl+P` (`cmd+P` on MacOS), which can be quicker than scrolling through the files pane if you have a lot of files.

> ![**Image 10:** *Opening existing files.*](https://codewithrepl.it/img/10-open-files.gif)

Author: ```suggestion
![**Image 12:** *Opening existing files.*](resources/10-open-files.png)
```

Reviewer: ```suggestion
![**Image 12:** *Opening existing files.*](resources/10-open-files.png)
```

If you've gotten used to either, you can emulate the experience in Repl.it by switching your keybinds. Go to the "Settings" tab and scroll down to where you can toggle between "default", "emacs" and "vim".

> ![**Image 12:** *Setting your keybinds to vim or emacs.*](https://codewithrepl.it/img/10-vim-emacs.png)

Reviewer: ```suggestion
![**Image 13:** *Setting your keybinds to vim or emacs.*](resources/10-vim-emacs.png)
```
